src/instructlab/model/serve.py

```diff
@@ -155,9 +164,29 @@ def serve(
             num_threads=num_threads,
         )
     elif backend == backends.VLLM:
+        # First Party
+        from instructlab.model.backends.vllm import contains_argument
+
         # Warn if unsupported backend parameters are passed
         warn_for_unsuported_backend_param(ctx)
 
+        if gpus:
+            # if `--tensor-parallel-size is included in ctx.args (click arguments after "--"),
+            # that value has precedence over the value of the --gpus flag.
+            if not contains_argument("--tensor-parallel-size", ctx.args):
+                ctx.args.extend(["--tensor-parallel-size", str(gpus)])
+            else:
+                logger.info(
+                    "'--gpus' flag used alongside '--tensor-parallel-size' flag in `ilab model serve`. Using value of the --tensor-parallel-size flag."
+                )
+
+            if contains_argument(
+                "--tensor-parallel-size", ctx.obj.config.serve.vllm.vllm_args
+            ):
+                logger.info(
+                    "'--gpus' flag used alongside '--tensor-parallel-size' in the vllm_args section of the config file. Using value of the --gpus flag."
+                )
+
         # Instantiate the vllm server
         if ctx.args:
```

Comment:
Does this line mean that when `--gpus` is passed, then arguments from `vllm_args` config section are ignored? (Even the arguments that have nothing to do with tensor size?) Is it intended here?

Reply:
I don't think `if ctx.args` has anything to do with --gpus directly.  This logic is replacing everything in vllm_args in config with what's after the --

I didn't actually realize it was doing this vs extending vllm_args.  Guessing that's the correct practice based on the author but should validate.

Reply:
Consider this: `ilab model serve --gpus=4`; also, in config file, `vllm_args` are set to `--random-option-to-vllm`. In this case,

1. `ctx.args` will start as empty list (because no `--` was passed.)
2. Then we `.extend` it in line 177.
3. Now, line 191 triggers, which will make the code use `--gpus=4` for `vllm_args`.
4. Line 195 is no longer triggered -> config file setting is ignored.

This feels wrong? We should probably merge the two sources of arguments (`--gpus=4` and the config `vllm_args`), not ignore the latter. I think this `if-else` here was designed so that when you pass `-- --more --options`, it allows you to override what is in the config file. But this is different?

Reply:
Ah, I see what you saw in the first place now!  Great catch!  I did not fully appreciate what was getting extended above.  I have the same instinct as you of asking why we are not updating vllm_args instead.

Reply:
I'd really love to see some targeted tests for this piece of code that would demonstrate all concerning permutations:

1. defaults.
2. gpus and no vllm-args.
3. vllm-args and no gpus.
4. gpus and vllm-args with just the tensor-parallel-size.
5. gpus and vllm-args with tensor-parallel-size only.
6. gpus and vllm-args with tensor-parallel-size AND some other options.
7. gpus and vllm-args with some other options but not tensor-parallel-size.

Perhaps there's more I missed.

Reply:
Dan is right. ctx.args is overwriting vllm_args for the server, instead of extending them. I would have expected extending `serve.vllm.vllm_args` to be the intended behavior. 

That doesn't have to do with the --gpus option though. Maybe we open up a separate PR for that?

Reply:
@booxter you're right. Given the current logic If the user specifies `--gpus` and ctx.args is empty then ctx.args is extended to have `--tensor-parallel-size` and `str(gpus)`.

Since `ctx.args` is not empty anymore, it gets assigned to `vllm_args` and that is what's passed into `vllm.Server()`, not `serve.vllm.vllm_args`. 

I'm going to make some changes to this PR and report back.

Reply:
The intended precedence for vllm argument setting I have after updating this PR is:
`ctx.args` > `--flag` > `serve.vllm.vllm_args`

where:
`ctx.args` is anything passed after `ilab model serve --`
`--flag` is  `--gpus` or any future flag for `ilab model serve` that would be an argument to vllm under the hood.
`serve.vllm.vllm_args` is what's in the config file.

@danmcp @booxter what do you think?

Reply:
I've checked the latest version, and I think the precedence is correct.

Reply:
The tests do a decent job of validating this precedence.  The only part which is a bit strange is gpus > tps in the config.  But tps > gpus in the command.  As long as we continue to output from the command which arg is winning, I think it's workable as is.

Reply:
> I'd really love to see some targeted tests for this piece of code that would demonstrate all concerning permutations:
> 
> 1. defaults.
I think this is covered
> 2. gpus and no vllm-args.
Covered
> 3. vllm-args and no gpus.
Covered
> 4. gpus and vllm-args with just the tensor-parallel-size.
Covered
> 5. gpus and vllm-args with tensor-parallel-size only.
Not sure how this is different than 4
> 6. gpus and vllm-args with tensor-parallel-size AND some other options.
Covered with ctx tps but not config.
> 7. gpus and vllm-args with some other options but not tensor-parallel-size.
Not yet covered
> 
> Perhaps there's more I missed.
I did cover several other combinations between the 4 places the value can be set.  If there are other important ones, we can either add them in this PR or open issues/PRs.